core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -116,29 +120,57 @@ public RecoverySourceHandler(final IndexShard shard, RecoveryTargetHandler recov
+                minRequiredSeqNo = SequenceNumbersService.UNASSIGNED_SEQ_NO;
+                maxRequiredSeqNo = SequenceNumbersService.UNASSIGNED_SEQ_NO;
 
-            try {
-                phase1(phase1Snapshot, translogView);
-            } catch (Exception e) {
-                throw new RecoveryEngineException(shard.shardId(), 1, "phase1 failed", e);
-            } finally {
+                final IndexCommit phase1Snapshot;
                 try {
-                    shard.releaseSnapshot(phase1Snapshot);
-                } catch (IOException ex) {
-                    logger.warn("releasing snapshot caused exception", ex);
+                    phase1Snapshot = shard.snapshotIndex(false);
+                } catch (Exception e) {
+                    IOUtils.closeWhileHandlingException(translogView);
+                    throw new RecoveryEngineException(shard.shardId(), 1, "Snapshot failed", e);
                 }
+
+                try {
+                    phase1(phase1Snapshot, translogView);
+                    prepareTargetForTranslog(translogView.totalOperations(), new Store.CommitId(phase1Snapshot));
+                } catch (Exception e) {
+                    throw new RecoveryEngineException(shard.shardId(), 1, "phase1 failed", e);
+                } finally {
+                    try {
+                        shard.releaseSnapshot(phase1Snapshot);
+                    } catch (IOException ex) {
+                        logger.warn("releasing snapshot caused exception", ex);
+                    }
+                }
+            } else {
+                // at the moment, we have no way to guarantee in advance that the translog view contains
+                // all operations from the request starting point and up. Instead we have to optimistically
+                // start recovering operations and verify at the end we sent what was needed. It's easy to check
+                // we saw the request start point and every operations above it up to a certain point. However
+                // at the current indexing edge there may be holes corresponding to operation that are currently
+                // being indexed and those are safe to recovery. Sadly we have no way of knowing which ops these are
+                // instead, we have to verify we have a nice consecutive range.
+                //
+                // nocommit: improving the translog allowing replay all ops from a seqNo up and check it contains a seqNo
+                // will allows us to remove this and also not fail and retry the recovery if we miss to find what we need
+                // but rather immediately start with a file based recovery
+                long currentMax = shard.seqNoStats().getMaxSeqNo();
+                logger.trace("{} waiting for all operations to complete up to [{}]", shard.shardId(), currentMax);
+                cancellableThreads.execute(() -> shard.waitForOpsToComplete(currentMax));
```

Comment:
no, it waits until all ops up to `currentMax` are finished locally. i.e., the local checkpoint has advanced beyond that point.